numpy/core/_add_newdocs.py

```diff
@@ -1333,6 +1330,25 @@
         this rule may result in the last element of `out` being greater
         than `stop`.
 
+    Warnings
+    --------
+    The length of the output might not be numerically stable.
+
+    Another stability issue is due to the internal implementation of `numpy.arange`.
+    The actual step value used to populate the array is
+    ``dtype(start + step) - dtype(start)`` and not `step`. Precision loss
+    can occur here, due to casting or due to using floating points when
+    `start` is much larger than `step`. This can lead to unexpected
+    behaviour. For example::
```

Comment:
I'm not sure about the wording here. As you mentioned previously in the discussion, it's hard to get the point across without getting too deeply into implementation details. IMO the clearest way to help users in the docstring is to provide examples of the bad behavior as below (though I think these examples are better described as arising from casting issues rather than numerical precision).

Reply:
I still want to include the actual ``dtype(start + step) - dtype(start)`` - I think it's more helpful for a lot of users than simply stating "this can be bad", although it might be confusing some (and that's why there are specific examples attached).

I do think that maybe swapping the sentence might be more clear - what about:

> Note that the step used might vary due to casting or due to precision loss. The internal step used is ``dtype(start + step) - dtype(start)`` and not `step`.